lib/streamlit/caching/cache_utils.py

```diff
@@ -228,6 +321,97 @@ def _show_cached_st_function_warning(
+and if the target DG is already known within the replay.
+
+Implicit invocation on a known DG -> Explicit invocation on that DG
+Implicit invocation on an unknown DG -> Rewrite as explicit invocation on main
+    This is situation 2 above, and the DG is a block entirely outside our function call,
+    so we interpret it as "put these elements within the current context", which is achieved
+    by invoking on main
+Explicit invocation on a known DG -> No change needed
+Explicit invocation on a known DG -> Raise an error
+    We have no way to identify the target DG, and it may not even be present in the
+    current script run, so the least surprising thing to do is raise an error
+
+"""
+
+
+class CacheMessagesCallStack(threading.local):
+    """A utility for storing messages generated by `st` commands called inside
+    a cached function.
+
+    Data is stored in a thread-local object, so it's safe to use an instance
+    of this class across multiple threads.
+    """
+
+    def __init__(self, cache_type: CacheType):
+        self._cached_message_stack: List[List[MsgData]] = []
+        self._seen_dg_stack: List[Set[str]] = []
+        self._most_recent_messages: List[MsgData] = []
+        self._cache_type = cache_type
+
+    def __repr__(self) -> str:
+        return util.repr_(self)
+
+    @contextlib.contextmanager
+    def calling_cached_function(self) -> Iterator[None]:
+        self._cached_message_stack.append([])
+        self._seen_dg_stack.append(set())
+        try:
+            yield
+        finally:
+            self._most_recent_messages = self._cached_message_stack.pop()
+            self._seen_dg_stack.pop()
+
+    def save_element_message(
+        self,
+        delta_type: str,
+        element_proto: Message,
+        invoked_dg_id: str,
+        used_dg_id: str,
+        returned_dg_id: str,
+    ) -> None:
+        """Record the element protobuf as having been produced during any currently
+        executing cached functions, so they can be replayed any time the function's
+        execution is skipped because they're in the cache.
+        """
+        id_to_save = self.get_dg_to_save_as_invoked(invoked_dg_id, used_dg_id)
+        for msgs in self._cached_message_stack:
+            msgs.append(
+                ElementMsgData(delta_type, element_proto, id_to_save, returned_dg_id)
+            )
+        for s in self._seen_dg_stack:
+            s.add(returned_dg_id)
+
+    def save_block_message(
+        self,
+        block_proto: Block,
+        invoked_dg_id: str,
+        used_dg_id: str,
+        returned_dg_id: str,
+    ) -> None:
+        id_to_save = self.get_dg_to_save_as_invoked(invoked_dg_id, used_dg_id)
+        for msgs in self._cached_message_stack:
+            msgs.append(BlockMsgData(block_proto, id_to_save, returned_dg_id))
+        for s in self._seen_dg_stack:
+            s.add(returned_dg_id)
+
+    def get_dg_to_save_as_invoked(self, invoked_id: str, acting_on_id: str) -> str:
```

Comment:
This needs a better name, especially since one of the parameters is described as "invoked"